File: storio-common-annotations-processor/src/main/java/com/pushtorefresh/storio/common/annotations/processor/introspection/StorIOCreatorMeta.java
package com.pushtorefresh.storio.common.annotations.processor.introspection;

import org.jetbrains.annotations.NotNull;

import java.lang.annotation.Annotation;

import javax.lang.model.element.Element;
import javax.lang.model.element.ExecutableElement;

public class StorIOCreatorMeta<CreatorAnnotation extends Annotation> {

    @NotNull
    public final Element enclosingElement;

    @NotNull
    public final ExecutableElement element;

    @NotNull
    public final CreatorAnnotation creatorAnnotation;

    public StorIOCreatorMeta(@NotNull Element enclosingElement,
                             @NotNull ExecutableElement element,
                             @NotNull CreatorAnnotation creatorAnnotation) {
        this.enclosingElement = enclosingElement;
        this.element = element;
        this.creatorAnnotation = creatorAnnotation;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        StorIOCreatorMeta<?> that = (StorIOCreatorMeta<?>) o;

        if (!enclosingElement.equals(that.enclosingElement)) return false;
        if (!element.equals(that.element)) return false;
        return creatorAnnotation.equals(that.creatorAnnotation);

    }

    @Override
    public int hashCode() {
        int result = enclosingElement.hashCode();
        result = 31 * result + element.hashCode();
        result = 31 * result + creatorAnnotation.hashCode();
        return result;
    }

    @Override
    public String toString() {
        return "StorIOCreatorMeta{" +
                "enclosingElement=" + enclosingElement +
                ", element=" + element +
                ", creatorAnnotation=" + creatorAnnotation +
                '}';
    }
}

File: storio-common-annotations-processor/src/main/kotlin/com/pushtorefresh/storio3/common/annotations/processor/introspection/StorIOCreatorMeta.kt
package com.pushtorefresh.storio3.common.annotations.processor.introspection

import javax.lang.model.element.Element
import javax.lang.model.element.ExecutableElement

open class StorIOCreatorMeta<out CreatorAnnotation : Annotation>(
        val enclosingElement: Element,
        val element: ExecutableElement,
        val creatorAnnotation: CreatorAnnotation) {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other?.javaClass != javaClass) return false

        other as StorIOCreatorMeta<*>

        if (enclosingElement != other.enclosingElement) return false
        if (element != other.element) return false
        if (creatorAnnotation != other.creatorAnnotation) return false

        return true
    }

    override fun hashCode(): Int {
        var result = enclosingElement.hashCode()
        result = 31 * result + element.hashCode()
        result = 31 * result + creatorAnnotation.hashCode()
        return result
    }

    override fun toString(): String = "StorIOCreatorMeta(enclosingElement=$enclosingElement, element=$element, creatorAnnotation=$creatorAnnotation)"

}